_posts/2014-11-13-cockpit-plugin-tutorial.md

```diff
@@ -78,92 +77,101 @@ Lets take a look at the pinger HTML, and see how it works.
-Next we attach a handler to the *Ping* button so that the `ping_run()` function is called when it is clicked.
+function ping_fail() {
+    result.style.color = "red";
+    result.innerHTML = "fail";
+}
+
+function ping_output(data) {
+    output.append(document.createTextNode(data));
+}
 
-```javascript
-$("#ping").on("click", ping_run);
+// Connect the button to starting the "ping" process
+button.addEventListener("click", ping_run);
 
+// Send a 'init' message.  This tells integration tests that we are ready to go
+cockpit.transport.wait(function() { });
+```
+
+First we get a bunch of variables pointing to the HTML elements we want to interact with. Next we attach a handler to the *Ping* button so that the `ping_run()` function is called when it is clicked.
+
+```js
 function ping_run() {
+}
+...
+button.addEventListener("click", ping_run);
 ```
 
-In the `ping_run()` function is where the magic happens. `cockpit.spawn` is a function, [documented here](https://cockpit-project.org/guide/latest/api-cockpit.html#latest-spawn) that lets you spawn processes on the server and interact with them via stdin and stdout. Here we spawn the `ping` command with some arguments:
+In the `ping_run()` function is where the magic happens. [cockpit.spawn](https://cockpit-project.org/guide/latest/cockpit-spawn.html) lets you spawn processes on the server and interact with them via stdin and stdout. Here we spawn the `ping` command with some arguments:
 
-```javascript
-    var proc = cockpit.spawn(["ping", "-c", "4", address.val()]);
+```js
+cockpit.spawn(["ping", "-c", "4", address.val()])
 ```
 
-In a web browser you cannot block and wait until a method call completes. Anything that doesn't happen instantaneously gets its results reported back to you by [means of callback handlers](https://cockpit-project.org/guide/latest/api-cockpit.html#latest-spawn-done). jQuery has a standard interface [called a promise](http://api.jquery.com/deferred.promise/). You add handlers by calling the `.done()` or `.fail()` methods and registering callbacks. `proc.stream()` registers a callback to be invoked whenever the process produces output.
+In a web browser you cannot block and wait until a method call completes. Anything that doesn't happen instantaneously gets its results reported back to you by [means of callback handlers](https://cockpit-project.org/guide/latest/cockpit-spawn.html#cockpit-spawn-then).
+
+JavaScript has a standard interface called a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). You add handlers by calling the `.then()` or `.catch()` methods and registering callbacks (Note: historically these have been called `.done()` and `.fail()`, but these should not be used any more in new code.) The `cockpit.spawn` specific `.stream()` handler registers a callback to be invoked whenever the process produces output.
```

Comment:
Should we have something around these paragraphs that have an important / warning alert that says it's important that you cannot depend on the browser session to be constant, due to _reasons_ (and list the reasons)?

It would help prevent issues with people coding pages that somehow expect 2-hour-long continuous browser sessions.

Reply:
That's a good idea, I added a warning.